### ISSUE TITLE
serving openpifpaf model for inference on TensorRT

### ISSUE THREAD

**Issue by udaygirish** [vita-epfl/openpifpaf#145]
Hi @svenkreiss,
I have been working on Pifpaf model and tried to deploy the model on Tensorrt Inference using ONNX Model format.
I have successfully deployed the model and I am getting 9 outputs.
But I'm not able to understand how to convert the 9 outputs(Pif-Paf fields) to get the key points and show it on the real-time image using Tensorrt.
Can you help with it??

**Comment by KevinHexin** [comment]
post-process module need you rewrite the python code in c++

**Comment by udaygirish** [comment]
Hi @KevinHexin 
Can you please elaborate on this, I didn't understand what you meant.

**Comment by udaygirish** [comment]
Hi,
This is the exact output which I am getting. 
(9 lists as an object )
![Screenshot from 2019-09-17 18-03-02](https://user-images.githubusercontent.com/18749512/65042032-a25e4980-d975-11e9-98f2-77854d63a7f2.png)

**Comment by svenkreiss** [comment]
Decoding is the process of going from fields to poses. It's Python code, not a neural network. Have a look at `openpifpaf/decoder/pifpaf.py`. If you just want to visualize the fields quickly, I have some code in the TypeScript demo here: https://github.com/vita-epfl/openpifpafwebdemo/blob/master/js/src/visualization.ts#L101
It's what generates the output of the web demo: https://vita-epfl.github.io/openpifpafwebdemo/

**Comment by udaygirish** [comment]
The issue was resolved using Openpifpaf preprocess codes. As I found out that internally the image is getting normalized so after sending the normalized array and post-processing the Tensor-RT output helped in getting the Keypoints. 
Thank you!!

**Comment by ll490187880** [comment]
> The issue was resolved using Openpifpaf preprocess codes. As I found out that internally the image is getting normalized so after sending the normalized array and post-processing the Tensor-RT output helped in getting the Keypoints.
> Thank you!!

Hello, @udaygirish  Can you share more details about your successful experiments? Thank you!

**Comment by BobDL** [comment]
> The issue was resolved using Openpifpaf preprocess codes. As I found out that internally the image is getting normalized so after sending the normalized array and post-processing the Tensor-RT output helped in getting the Keypoints.
> Thank you!!
Does the Tensorrt great impove the speed?
Thanks

**Comment by udaygirish** [comment]
First I tried to understand individual code blocks in the repo and I figured out which input is needed for the TF graph and the respective output obtained from it. From that, I understood the preprocessing and post-processing steps and also came to know that the ONNX model generated by the code with this repo contains only the part which gives you fields. So, that is the exact thing required by a Tensor RT engine as TRT is only responsible for speeding up the graph-based calculations using NVIDIA libraries. Then I made a code block to preprocess the input and passed it to TRT endpoint and then post-processed the output from it which resulted in the output I expected.

Detailed execution of TRT on the different types of models is there in the NVIDIA Tensor RT documentation. 
Yeah, it did a great improvement in terms of both latency and accuracy in case of pose estimation.
(It depends on dataset too)

**Comment by bfortuner** [comment]
Oh that's cool! @udaygirish do you have any sample code?

**Comment by anuar12** [comment]
@udaygirish how did you convert the ONNX model to TensorRT? Can you please elaborate or share some code? Thank you!

**Comment by nkhdiscovery** [comment]
I also have tried converting ONNX model to trt using onnx2trt (no extra parameters or changes), the model converts fine but the outputs seem wrong (I have tested my visualization code on ONNX's output - converting fields to annotations - keypoints - boxes - and it works fine so I am almost sure the only thing not working properly is the trt engine exported)

Anyone can give me some information on how could get it to work with TensorRT? 
I am using version 0.10.0 of repo (for some reason) on jetson TX2 

@mdenna-nviso  @KevinHexin seems you guys have some experience with tensorrt, let me know how I may debug this and make sure the trt engine converted from ONNX is converted correctly. 

@svenkreiss I may be able to solve my problem with using #188 (not tested yet) but I am curious why this way the tensorrt inference does not work properly. I am eager to contribute, let me know what I can do. 

By the way, I compared the fields, they seem different, and also while decoding the output, in `PifSeeds` class, I printed the vavlues in `fill` class and saw values differ a lot in comparison to original inference on the cpu with the same version (0.10.0).

**Comment by svenkreiss** [comment]
Hi @nkhdiscovery , I'd like to help. The main reason for the backwards-incompatible change in the dev branch is the simplification of the inference graph for engines like tensorrt. We have found different reasons in different scenarios why the conversion might fail. The dev branch finally has a test case that checks that the ONNX graph works: https://github.com/vita-epfl/openpifpaf/blob/10adda7a20c84f3a5fd703618fb893ac488367f4/tests/test_onnx_export.py#L34
Such a test, which guarantees a working ONNX model, did not exist before.
There is also the open #281 which is worth having a look at: it suggests a few further changes and mentions that TensorRT only likes ResNets so far.

**Comment by nkhdiscovery** [comment]
> Hi @nkhdiscovery , I'd like to help. The main reason for the backwards-incompatible change in the dev branch is the simplification of the inference graph for engines like tensorrt. We have found different reasons in different scenarios why the conversion might fail. The dev branch finally has a test case that checks that the ONNX graph works:
> 
> https://github.com/vita-epfl/openpifpaf/blob/10adda7a20c84f3a5fd703618fb893ac488367f4/tests/test_onnx_export.py#L34
> 
> Such a test, which guarantees a working ONNX model, did not exist before.
Hi @svenkreiss , thanks for your response.
As ONNX converts fine with good output on 0.10.0 version, and TensorRT also converts fine with no errors, I do not think that this is the case with that version of the code. Heads are fine and ONNX output is also fine, decodes fine and works great. The problem is that the TensorRT engine created from that ONNX gives unreasonable outputs, and by unreasonable, I mean after decoding it does not provide sane bboxes, for example. Although, seems it varies for images with different number of people in it, so it seems also somehow working. The problem is usually that w, h for bounding boxes becomes zero and their x, y also seem to be in some wrong places.

This is great that we have a test for ONNX, the problem is now how to make sure that the TensorRT engine is converted fine, having correct weights (shapes are correct, and heads are fine). If we come to understand how one should convert the ONNX to TensorRT and inference using it, we then will be able to write a test for TensorRT too, which will make even more people happy :D 

> There is also the open #281 which is worth having a look at: it suggests a few further changes and mentions that TensorRT only likes ResNets so far.
Yeah, I am working with ResNet50 (and then maybe it's the reason it converts fine). If we can overcome the issue of inference we can then move to test other networks as well. 

Time to say that your work is great, and it worth having all these tools in it (converting/testing to ONNX,TensorRT, maybe someday tflite for edgetpu ... ). Thanks for your great work.

**Comment by nkhdiscovery** [comment]
Once again, this is different from #281 as the problem is not to convert torch2trt, the problem is that conversion with onnx2trt is successful and inference with that trt engine gives no errors, but gives bad results.

**Comment by svenkreiss** [comment]
A test with TensorRT would be great. We need this test to run on every commit like all the other tests. When I looked into it before, it seemed that you always need a machine with an Nvidia gpu to run TensorRT. Do you know whether that is still the case? Our test infrastructure (TravisCI) does not support GPUs.

**Comment by nkhdiscovery** [comment]
> A test with TensorRT would be great. We need this test to run on every commit like all the other tests. 
That needs the problem to be solved at once, and that is why started this conversation here :D 
I need help getting it to work in the first place. 

> When I looked into it before, it seemed that you always need a machine with an Nvidia gpu to run TensorRT. Do you know whether that is still the case? Our test infrastructure (TravisCI) does not support GPUs.
Yeah, that is still the case, TensorRT is built on CUDA so if we solve it, you may only include the tools in the repo, but not include the test on the CI.

**Comment by thancaocuong** [comment]
@nkhdiscovery conversion from onnx to tensorRT does not guarantee that the output will be the same as the first one. I've found that the best way to do conversion is that we need to write the network by TensorRT Layers, and then loads weights from pytorch checkpoint. It's the same way as Yolo did. For some details, you could follow this repo https://github.com/wang-xinyu/tensorrtx

**Comment by nkhdiscovery** [comment]
> @nkhdiscovery conversion from onnx to tensorRT does not guarantee that the output will be the same as the first one. I've found that the best way to do conversion is that we need to write the network by TensorRT Layers, and then loads weights from pytorch checkpoint. It's the same way as Yolo did. For some details, you could follow this repo https://github.com/wang-xinyu/tensorrtx

Sure, I have this experience with TensorRT and for sure the optimized version does not have "exactly" the same output, but it still makes sense. Mine does not.

I accept that your suggestion maybe the best way, but conversion from ONNX still has to work.

**Comment by nkhdiscovery** [comment]
Is there a way that I can visualize Pifs and Pafs independently so I can compare the output between the trt and original model? 
One thing hindering me is this, I can not decide if Pifs and Pafs make sense or not.

**Comment by thancaocuong** [comment]
@nkhdiscovery  the only way I do now is that I compare the max  abs of L1 for output from tensorRT and onnx, pytorch

**Comment by undefined-references** [comment]
I'm exited to announce that we have generated a TensorRT engine on Jetson TX2 from openpifpaf exported ONNX (version 0.10.0) and run inference on [Neuralet](https://github.com/neuralet/neuralet/tree/master/applications/pose-estimation-tensorrt). We have converted ONNX model using [ONNX-TensorRT](https://github.com/onnx/onnx-tensorrt) and implemented inference scripts by using decoder modules from openpifpaf. 
You can find more detailed explanations [here](https://neuralet.com/article/pose-estimation-on-nvidia-jetson-platforms-using-openpifpaf/) on Neuralet website.   
@thancaocuong @nkhdiscovery @svenkreiss @anuar12 @ll490187880

**Comment by svenkreiss** [comment]
Fantastic news @undefined-references

**Comment by thancaocuong** [comment]
Great news @undefined-references
